src/pip/_internal/resolution/resolvelib/resolver.py

```diff
@@ -186,8 +186,7 @@ def get_installation_order(
 
         graph = self._result.graph
         weights = get_topological_weights(
-            graph,
-            expected_node_count=len(self._result.mapping) + 1,
+            graph, requirement_keys=set(req_set.requirements.keys())
```

Comment:
This gives a mypy error:

```
src/pip/_internal/resolution/resolvelib/resolver.py:280: error: Argument 1 to
"len" has incompatible type "Container[str]"; expected "Sized"  [arg-type]
        assert len(weights) == len(requirement_keys)
```

So I tried `Sized[str]` but `Sized` does not expect any arguments.

I don't know enough about mypy and type hints to solve this in a reasonable time. `pip` is the first project where I see them, so I am hitting a learning curve here. Learning this is probably time well spent, and I can imagine it saves the maintainers time, by keeping the source code in a sane state. But for now it takes too much time. Ah well, probably best to ignore this mini-rant.

But apart from type hinting stress, I would advocate that we can keep the argument a Set:

- `get_topological_weights` checks lots of times if a key is in the requirement_keys, which should be faster when this is a Set already.
- When we have a Set as argument, instead of comparing the lengths of `weights` and `requirement_keys`, we can compare the actual contents, like this:

```
difference = set(weights.keys()).difference(requirement_keys)
assert not difference, difference
```

If the assertion fails, we get a more informative error message.
I will push a fix to my PR.

Reply:
`Collection[str]` should be the simplies fix.

Reply:
Ah, I checked, and this would have indeed worked.